File: typescript/src/days/day05.ts
// Import NodeJS helpers
import fs from 'fs';

// Import types
import InputFile from '../types/InputFile';

// ------------------ Types ------------------ //

type SupplyStacks = string[][];

type CraneMove = {
  numCrates: number,
  from: number,
  to: number,
};

// ----------------- Helpers ----------------- //

const splitInput = async (filename: InputFile) => {
  const inputFile = await fs.promises.readFile(filename);
  const [revCrates, moves] = inputFile
    .toString()
    .split('\n\n')
    .map((str) => str.split('\n'))
    .map((list) => { list.pop(); return list; });

  const crates = revCrates.reverse();

  return {
    crates,
    moves,
  };
};

const parseRow = (row: string) => {
  const crateRow: string[] = [];
  for (let i = 1; i < row.length; i += 4) {
    crateRow.push(row.charAt(i));
  }

  return crateRow;
};

const initSupplyStacks = (rows: string[]): SupplyStacks => {
  const numRows = (rows[0].length + 1) / 4;
  const supplyStacks: SupplyStacks = Array(numRows)
    .fill('')
    .map((s) => [s]);

  rows
    .map(parseRow)
    .forEach((row) => (
      row.forEach((crate, col) => {
        if (crate === ' ') return;
        supplyStacks[col].push(crate);
      })
    ));

  return supplyStacks;
};

const parseMove = (moveStr: string): CraneMove => {
  const moveList = moveStr
    .split(' ')
    .map((str) => Number.parseInt(str, 10))
    .filter((num) => !Number.isNaN(num));

  return {
    numCrates: moveList[0],
    from: moveList[1] - 1,
    to: moveList[2] - 1,
  };
};

const crateMover9000 = (supplyStacks: SupplyStacks, move: CraneMove): SupplyStacks => {
  const newSupplyStacks = supplyStacks;
  const {
    numCrates,
    from,
    to,
  } = move;

  for (let i = 0; i < numCrates; i++) {
    newSupplyStacks[to].push(newSupplyStacks[from].pop());
  }

  return newSupplyStacks;
};

const crateMover9001 = (supplyStacks: SupplyStacks, move: CraneMove): SupplyStacks => {
  const newSupplyStacks = supplyStacks;
  const {
    numCrates,
    from,
    to,
  } = move;

  const cratesToMove = newSupplyStacks[from].splice(-numCrates, numCrates);

  newSupplyStacks[to] = newSupplyStacks[to].concat(cratesToMove);

  return newSupplyStacks;
};

// ------------- Main Functions -------------- //

/**
 * Day 5 Pt1 of Advent of Code 2022!
 * Results according to the Crate Mover 9000.
 * @author Benedikt Arnarsson
 */
const pt1 = async (filename: InputFile) => {
  const {
    crates,
    moves,
  } = await splitInput(filename);

  const supplyStacks = initSupplyStacks(crates);

  return moves
    .map(parseMove)
    .reduce(crateMover9000, supplyStacks)
    .map((col) => col.pop())
    .join('');
};

/**
 * Day 5 Pt2 of Advent of Code 2022!
 * Results according to the Crate Mover 9001.
 * @author Benedikt Arnarsson
 */
const pt2 = async (filename: InputFile) => {
  const {
    crates,
    moves,
  } = await splitInput(filename);

  const supplyStacks = initSupplyStacks(crates);

  return moves
    .map(parseMove)
    .reduce(crateMover9001, supplyStacks)
    .map((col) => col.pop())
    .join('');
};

const day05 = [pt1, pt2];

export default day05;

File: python/days/day05.py
"""Day 5 of AOC 2022: Supply Stacks"""

# -------------- Public Functions ------------- #

def day05_pt1(filename: str) -> str:
    """Day 5, part 1: Crate Mover 9000"""
    init_crates, moves = _split_input(filename)

    supply_stacks = SupplyStacks(init_crates)

    for move in moves:
        supply_stacks.crate_mover_9000(move)

    return supply_stacks.top_crates()

def day05_pt2(filename: str) -> str:
    """Day 5, part 2: Crate Mover 9001"""
    init_crates, moves = _split_input(filename)

    supply_stacks = SupplyStacks(init_crates)

    for move in moves:
        supply_stacks.crate_mover_9001(move)

    return supply_stacks.top_crates()


# ----------------- Classes ----------------- #

class SupplyStacks:
    """The stacks of crates that we are moving around."""
    crates: list[list[str]]

    def __init__(self, rows: list[str]):
        # TODO:
        num_cols = (len(rows[0]) + 1)//4
        self.crates = [[] for _ in range(num_cols)]
        for row in rows:
            self.add_row(row)

    def add_row(self, row: str):
        """Adds a row of crates to the top of the stacks."""
        crate_list = [row[i:i+4][1] for i in range(0, len(row), 4)]
        for i, crate in enumerate(crate_list):
            if crate != ' ':
                self.crates[i].append(crate)

    def crate_mover_9000(self, move: str):
        """Move crates as specified by the Crate Mover 9000 manual."""
        num_crates, from_stack, to_stack = _parse_move(move)
        for _ in range(num_crates):
            self.crates[to_stack].append(self.crates[from_stack].pop())

    def crate_mover_9001(self, move: str):
        """Move crates as specified by the Crate Mover 9001 manual."""
        num_crates, from_stack, to_stack = _parse_move(move)

        self.crates[to_stack] += self.crates[from_stack][-num_crates:]
        self.crates[from_stack] = self.crates[from_stack][:-num_crates]

    def top_crates(self) -> str:
        """Get a string of the set of crates at the top of each stack."""
        return ''.join([col.pop() for col in self.crates])

# ----------------- Constants ----------------- #


# ------------- Private Functions ------------- #

def _split_input(filename: str) -> list[list[str]]:
    with open(filename, 'r', encoding='utf-8') as input_file:
        lines = input_file.read().split('\n\n')

    [init_crates, moves] = list(map(
        lambda line: line.split('\n')[:-1],
        lines,
    ))

    init_crates = list(reversed(init_crates))

    return init_crates, moves

def _parse_move(move_str: str) -> tuple[int, int, int]:
    [num_crates, from_stack, to_stack] = list(map(int, move_str.split(' ')[1::2]))
    return num_crates, from_stack - 1, to_stack - 1


# ------------------- Script ------------------ #

if __name__ == '__main__':
    INPUT_FILENAME = '../inputs/05.txt'
    print(f'Day 5, pt1: {day05_pt1(INPUT_FILENAME)}')
    print(f'Day 5, pt2: {day05_pt2(INPUT_FILENAME)}')
